### ISSUE TITLE
Turtle writer collection blank edge case

### ISSUE THREAD

**Issue by langsamu** [dotnetrdf/dotnetrdf#279]
### Summary
The [Turtle writer](https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Writing/CompressingTurtleWriter.cs) on high compression level incorrectly serialises a related RDF collection with a single item.

### Details
For the following topology
![image](https://user-images.githubusercontent.com/240741/70341671-62f41a80-184b-11ea-92b7-abb37f1750c7.png)
[`CompressionLevel.Medium`](https://github.com/dotnetrdf/dotnetrdf/blob/6bb86a68624759fb09c02293b63b8f0e65b2c2cb/Libraries/dotNetRDF/Writing/WriterUtilities.cs#L80) produces this correct Turtle
```ttl
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.

_:autos1 a _:autos2.
_:autos2 rdf:first _:autos3;
         rdf:rest rdf:nil.
_:autos3 a _:autos4.
```
while [`CompressionLevel.More`](https://github.com/dotnetrdf/dotnetrdf/blob/6bb86a68624759fb09c02293b63b8f0e65b2c2cb/Libraries/dotNetRDF/Writing/WriterUtilities.cs#L84) produces this incurrect Turtle
```ttl
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.

[] a ([]).
[] a [].
```
![image](https://user-images.githubusercontent.com/240741/70341679-67203800-184b-11ea-9d88-5448d4e07393.png)

### Code to reproduce
```cs
var g = new Graph();
g.LoadFromString("[ a ( [ a [] ] ) ] .", new TurtleParser());

var mediumString = StringWriter.Write(g, new CompressingTurtleWriter(WriterCompressionLevel.Medium));
var moreString = StringWriter.Write(g, new CompressingTurtleWriter(WriterCompressionLevel.More));
            
Console.WriteLine(mediumString);
Console.WriteLine();
Console.WriteLine(moreString);

var medium = new Graph();
medium.LoadFromString(mediumString);

var more = new Graph();
more.LoadFromString(moreString);

Assert.AreEqual(g, medium); // Succeeds
Assert.AreEqual(g, more); // Fails
```

### Working cases

These roundrip correctly.

| Code | Topology |
| --- | --- |
| `[ a ( [ a [ a [] ] ] ) ] .` | ![image](https://user-images.githubusercontent.com/240741/70342783-a0f23e00-184d-11ea-8d6d-e0912b8cdb3e.png) |
| `[ a ( [ a [], [] ] ) ] .` | ![image](https://user-images.githubusercontent.com/240741/70342803-a94a7900-184d-11ea-9776-38c827d8d616.png) |
| `[ a ( [ a [] ] [] ) ] .` | ![image](https://user-images.githubusercontent.com/240741/70343170-7ce32c80-184e-11ea-8813-7c8b80d92ce4.png) |

**Comment by kal** [comment]
Thanks for the report @langsamu.
I took a look at the code that determines how to compress blank nodes and its pretty complex. I think I found the right fix for this edge case. It definitely writes out a correct representation of the graph now and all of the other cases you have here as known-good also roundtrip correctly, although the last one does not compress as expected.

I've pushed a fix to a branch and I'd appreciate it if you get chance to take a look and verify it. I'll create a PR for that right now...
